File: backend/app/seed.py
from sqlalchemy.orm import Session

from app.models import Framework, Guide, Model, Tool


def seed_database(db: Session) -> None:
    if db.query(Tool).count() > 0:
        return

    tools = [
        Tool(
            slug="ollama",
            name="Ollama",
            tagline="Run open models locally with one command",
            description=(
                "Ollama packages LLMs into a simple CLI and REST API. Pull a model with "
                "`ollama pull llama3`, then chat from the terminal or hit localhost:11434. "
                "It's the fastest way for developers to prototype local AI without managing GPUs manually."
            ),
            category="runtime",
            website_url="https://ollama.com",
            github_url="https://github.com/ollama/ollama",
            tags="cli,api,macos,linux,windows,beginner-friendly",
            difficulty="beginner",
        ),
        Tool(
            slug="lm-studio",
            name="LM Studio",
            tagline="Desktop GUI for discovering and running local models",
            description=(
                "LM Studio provides a polished desktop app to browse Hugging Face models, "
                "download GGUF weights, and chat or expose an OpenAI-compatible local server. "
                "Great for developers who prefer a visual workflow before scripting."
            ),
            category="desktop",
            website_url="https://lmstudio.ai",
            github_url=None,
            tags="gui,gguf,openai-compatible,desktop",
            difficulty="beginner",
        ),
        Tool(
            slug="llama-cpp",
            name="llama.cpp",
            tagline="High-performance C++ inference engine",
            description=(
                "llama.cpp is the foundational inference project behind much of the local AI ecosystem. "
                "It runs quantized models efficiently on CPU and GPU, powers many wrappers, and is ideal "
                "when you need maximum control or embedded deployments."
            ),
            category="inference",
            website_url="https://github.com/ggerganov/llama.cpp",
            github_url="https://github.com/ggerganov/llama.cpp",
            tags="c++,gguf,cpu,gpu,embedded",
            difficulty="intermediate",
        ),
        Tool(
            slug="localai",
            name="LocalAI",
            tagline="Drop-in OpenAI API replacement that runs locally",
            description=(
                "LocalAI exposes endpoints compatible with the OpenAI API while running models on your hardware. "
                "Swap your app's base URL and keep existing SDK code — useful for testing privacy-sensitive flows."
            ),
            category="api-server",
            website_url="https://localai.io",
            github_url="https://github.com/mudler/LocalAI",
            tags="openai-compatible,docker,self-hosted,api",
            difficulty="intermediate",
        ),
        Tool(
            slug="text-generation-webui",
            name="Text Generation WebUI",
            tagline="Gradio-based web UI for experimenting with models",
            description=(
                "oobabooga's text-generation-webui is a feature-rich browser interface supporting many backends "
                "(llama.cpp, Transformers, ExLlama). Developers use it to benchmark prompts and extensions before production."
            ),
            category="desktop",
            website_url="https://github.com/oobabooga/text-generation-webui",
            github_url="https://github.com/oobabooga/text-generation-webui",
            tags="gradio,extensions,experimentation,python",
            difficulty="intermediate",
        ),
        Tool(
            slug="vllm",
            name="vLLM",
            tagline="Production-grade LLM serving with PagedAttention",
            description=(
                "vLLM is built for high-throughput serving on NVIDIA GPUs. If you're moving from a laptop prototype "
                "to an internal API that handles concurrent requests, vLLM is a common next step."
            ),
            category="serving",
            website_url="https://vllm.ai",
            github_url="https://github.com/vllm-project/vllm",
            tags="gpu,serving,throughput,production",
            difficulty="advanced",
        ),
    ]

    models = [
        Model(
            slug="llama-3-8b",
            name="Llama 3 8B Instruct",
            provider="Meta",
            parameter_size="8B",
            context_length="8K",
            description=(
                "A strong general-purpose instruct model that runs on consumer GPUs with quantization. "
                "Good default for chatbots, summarization, and code assistance prototypes."
            ),
            use_cases="chat,summarization,code-assist",
            license="Llama 3 Community License",
            tags="instruct,meta,general-purpose",
            min_vram_gb=6,
        ),
        Model(
            slug="mistral-7b",
            name="Mistral 7B Instruct",
            provider="Mistral AI",
            parameter_size="7B",
            context_length="32K",
            description=(
                "Efficient 7B model with long context support. Popular for RAG pipelines and agent tooling "
                "where you need a balance of speed and quality on modest hardware."
            ),
            use_cases="rag,agents,chat",
            license="Apache 2.0",
            tags="instruct,mistral,long-context",
            min_vram_gb=6,
        ),
        Model(
            slug="phi-3-mini",
            name="Phi-3 Mini 4K Instruct",
            provider="Microsoft",
            parameter_size="3.8B",
            context_length="4K",
            description=(
                "Small but capable model designed for edge and laptop deployment. Excellent when VRAM is limited "
                "or you need fast iteration loops during development."
            ),
            use_cases="edge,fast-prototyping,chat",
            license="MIT",
            tags="small,efficient,microsoft",
            min_vram_gb=4,
        ),
        Model(
            slug="codellama-13b",
            name="Code Llama 13B Instruct",
            provider="Meta",
            parameter_size="13B",
            context_length="16K",
            description=(
                "Specialized for code generation, explanation, and infilling. Use when building dev tools, "
                "local copilots, or documentation assistants inside your IDE workflow."
            ),
            use_cases="code-generation,infilling,dev-tools",
            license="Llama 2 Community License",
            tags="code,meta,instruct",
            min_vram_gb=10,
        ),
        Model(
            slug="qwen2-5-7b",
            name="Qwen2.5 7B Instruct",
            provider="Alibaba",
            parameter_size="7B",
            context_length="128K",
            description=(
                "Multilingual model with very long context. Strong choice for document Q&A and mixed-language "
                "support in local deployments."
            ),
            use_cases="multilingual,rag,long-documents",
            license="Apache 2.0",
            tags="multilingual,long-context,qwen",
            min_vram_gb=6,
        ),
        Model(
            slug="deepseek-r1-distill",
            name="DeepSeek R1 Distill 7B",
            provider="DeepSeek",
            parameter_size="7B",
            context_length="32K",
            description=(
                "Reasoning-focused distilled model that fits consumer hardware. Useful for experimenting with "
                "chain-of-thought style apps without cloud API costs."
            ),
            use_cases="reasoning,math,agents",
            license="MIT",
            tags="reasoning,distilled,chain-of-thought",
            min_vram_gb=8,
        ),
    ]

    frameworks = [
        Framework(
            slug="langchain",
            name="LangChain",
            language="Python",
            description=(
                "Composable framework for building LLM apps: chains, agents, retrievers, and tool calling. "
                "Supports local models via Ollama, llama.cpp, and OpenAI-compatible servers."
            ),
            website_url="https://www.langchain.com",
            github_url="https://github.com/langchain-ai/langchain",
            tags="agents,rag,tools,python",
        ),
        Framework(
            slug="llamaindex",
            name="LlamaIndex",
            language="Python",
            description=(
                "Data framework focused on connecting LLMs to your documents. Ideal for local RAG pipelines "
                "where embeddings and retrieval run entirely on your machine."
            ),
            website_url="https://www.llamaindex.ai",
            github_url="https://github.com/run-llama/llama_index",
            tags="rag,embeddings,indexing,python",
        ),
        Framework(
            slug="open-webui",
            name="Open WebUI",
            language="TypeScript",
            description=(
                "Self-hosted ChatGPT-style interface that plugs into Ollama and OpenAI-compatible backends. "
                "Teams use it as an internal chat portal over local models."
            ),
            website_url="https://openwebui.com",
            github_url="https://github.com/open-webui/open-webui",
            tags="ui,ollama,self-hosted,typescript",
        ),
        Framework(
            slug="semantic-kernel",
            name="Semantic Kernel",
            language="C#",
            description=(
                "Microsoft's SDK for orchestrating AI plugins and planners. Works with local endpoints "
                "through OpenAI-compatible adapters — useful in .NET shops."
            ),
            website_url="https://learn.microsoft.com/semantic-kernel",
            github_url="https://github.com/microsoft/semantic-kernel",
            tags="dotnet,plugins,orchestration",
        ),
        Framework(
            slug="haystack",
            name="Haystack",
            language="Python",
            description=(
                "Pipeline framework for search, QA, and generative AI. Build production RAG with local "
                "embedders and generators in a modular graph."
            ),
            website_url="https://haystack.deepset.ai",
            github_url="https://github.com/deepset-ai/haystack",
            tags="rag,pipelines,search,python",
        ),
    ]

    guides = [
        Guide(
            slug="why-local-ai",
            title="Why Local AI Matters for Developers",
            summary="Privacy, cost control, offline dev, and faster iteration — the case for running models on your machine.",
            content="""# Why Local AI Matters for Developers

Running models locally isn't just a hobbyist pursuit — it's a practical development strategy.

## Key benefits

**Privacy & compliance** — User data never leaves your network. Healthcare, legal, and enterprise prototypes often require this from day one.

**Predictable costs** — Cloud inference bills scale with usage. Local hardware is a fixed cost; great for high-volume dev/test loops.

**Offline & low-latency** — No network round trips. Iterate on prompts and agents on a plane or in a locked-down environment.

**Full control** — Pick the model, quantization, context length, and tooling. No vendor rate limits during development.

## When to start local

- Building RAG over sensitive documents
- Prototyping agents with tool calling
- Learning how transformers behave under the hood
- Creating demos without API keys

## When to move to cloud

Production scale, largest frontier models, or managed fine-tuning may still need cloud — but local dev shrinks the gap dramatically.

**Next:** Set up Ollama and run your first model in under five minutes.""",
            level="beginner",
            reading_time_minutes=4,
            tags="introduction,overview,privacy",
            order_index=1,
        ),
        Guide(
            slug="first-model-ollama",
            title="Run Your First Local Model with Ollama",
            summary="Install Ollama, pull a model, and call it from curl or Python in minutes.",
            content="""# Run Your First Local Model with Ollama

## 1. Install

```bash
curl -fsSL https://ollama.com/install.sh | sh
```

Or download the macOS/Windows app from [ollama.com](https://ollama.com).

## 2. Pull a model

```bash
ollama pull llama3
```

This downloads weights optimized for your hardware.

## 3. Chat in the terminal

```bash
ollama run llama3 "Explain recursion to a junior dev"
```

## 4. Hit the REST API

```bash
curl http://localhost:11434/api/generate -d '{
  "model": "llama3",
  "prompt": "Write a FastAPI hello world",
  "stream": false
}'
```

## 5. Use from Python

```python
import requests

resp = requests.post(
    "http://localhost:11434/api/generate",
    json={"model": "llama3", "prompt": "Hello", "stream": False},
)
print(resp.json()["response"])
```

You now have a local inference server — no API key required.""",
            level="beginner",
            reading_time_minutes=6,
            tags="ollama,quickstart,api",
            order_index=2,
        ),
        Guide(
            slug="choosing-hardware",
            title="Choosing Hardware for Local AI",
            summary="VRAM, RAM, and CPU tradeoffs — what you need to run different model sizes.",
            content="""# Choosing Hardware for Local AI

## VRAM is usually the bottleneck

| Model size (Q4 quant) | Approx VRAM |
|-----------------------|-------------|
| 3B                    | 4 GB        |
| 7–8B                  | 6–8 GB      |
| 13B                   | 10–12 GB    |
| 70B                   | 48 GB+      |

Quantization (GGUF Q4_K_M, etc.) dramatically reduces memory at some quality cost.

## No GPU?

**llama.cpp** runs on CPU with enough RAM — slower but workable for small models and scripting.

## Apple Silicon

Unified memory means a 16–32 GB M-series Mac can run 7B–13B models comfortably via Metal acceleration in Ollama and llama.cpp.

## Development vs production

- **Dev laptop:** 7B instruct models for fast iteration
- **Workstation / server:** 13B+ or vLLM for team APIs
- **CI:** CPU-only smoke tests on tiny models

Match the model to your hardware using our catalog filters.""",
            level="beginner",
            reading_time_minutes=5,
            tags="hardware,vram,gpu,cpu",
            order_index=3,
        ),
        Guide(
            slug="rag-basics",
            title="RAG Basics: Chat With Your Docs Locally",
            summary="Embeddings, chunking, and retrieval — build a private document Q&A pipeline.",
            content="""# RAG Basics: Chat With Your Docs Locally

Retrieval-Augmented Generation (RAG) grounds the model in *your* data.

## Pipeline overview

1. **Load** documents (PDF, Markdown, code)
2. **Chunk** into overlapping segments
3. **Embed** each chunk with a local embedding model
4. **Store** vectors in a local DB (Chroma, LanceDB, sqlite-vss)
5. **Retrieve** relevant chunks at query time
6. **Generate** an answer with context in the prompt

## Minimal stack

- **Ollama** for chat + embeddings (`nomic-embed-text`)
- **LlamaIndex** or **LangChain** for orchestration
- **Chroma** for vector storage

## Privacy win

Documents and embeddings never leave your machine — ideal for internal wikis and codebases.

## Common pitfalls

- Chunks too large → noisy retrieval
- No metadata filters → wrong doc versions
- Skipping evaluation → hallucinations look plausible

Start small: one directory of Markdown files and a 20-question eval set.""",
            level="intermediate",
            reading_time_minutes=8,
            tags="rag,embeddings,llamaindex,privacy",
            order_index=4,
        ),
        Guide(
            slug="openai-compatible-local",
            title="Swap Cloud APIs for Local OpenAI-Compatible Servers",
            summary="Point existing SDKs at LocalAI or Ollama's compatibility layer with minimal code changes.",
            content="""# Swap Cloud APIs for Local OpenAI-Compatible Servers

Many local tools expose an OpenAI-compatible `/v1/chat/completions` endpoint.

## Ollama OpenAI compatibility

```bash
ollama pull llama3
```

```python
from openai import OpenAI

client = OpenAI(
    base_url="http://localhost:11434/v1",
    api_key="ollama",  # required but unused
)

completion = client.chat.completions.create(
    model="llama3",
    messages=[{"role": "user", "content": "Hello!"}],
)
print(completion.choices[0].message.content)
```

## Why this matters

- Reuse existing OpenAI SDK code in tests
- Toggle `base_url` between local and cloud via env vars
- Run integration tests in CI with tiny models

## LocalAI & vLLM

**LocalAI** and **vLLM** also offer compatible APIs for heavier deployments.

```bash
export OPENAI_BASE_URL=http://localhost:8080/v1
export OPENAI_API_KEY=local
```

Your app stays the same; only the inference backend changes.""",
            level="intermediate",
            reading_time_minutes=7,
            tags="openai-compatible,localai,ollama,migration",
            order_index=5,
        ),
    ]

    db.add_all(tools + models + frameworks + guides)
    db.commit()

File: prisma/seed.ts
import { PrismaClient } from "@prisma/client";
import { hash } from "bcryptjs";

const prisma = new PrismaClient();

const models = [
  {
    slug: "llama-3-1-8b",
    name: "Llama 3.1 8B",
    vendor: "Meta",
    description:
      "A strong general-purpose open model that runs well on a modern laptop GPU or Apple Silicon. Great first local chat model for coding assistants and RAG experiments.",
    sizeParams: "8B",
    license: "Llama 3.1 Community",
    contextLength: 128000,
    tags: JSON.stringify(["chat", "coding", "rag", "general"]),
    hardwareMin: "16GB RAM + GPU/Metal",
    ollamaName: "llama3.1:8b",
    hfUrl: "https://huggingface.co/meta-llama/Llama-3.1-8B-Instruct",
    useCases: JSON.stringify(["chatbots", "code assist", "summarization"]),
  },
  {
    slug: "mistral-7b-instruct",
    name: "Mistral 7B Instruct",
    vendor: "Mistral AI",
    description:
      "Efficient instruction-tuned model with solid reasoning for its size. Excellent latency-to-quality tradeoff for local APIs and CLI agents.",
    sizeParams: "7B",
    license: "Apache 2.0",
    contextLength: 32768,
    tags: JSON.stringify(["chat", "instruct", "efficient"]),
    hardwareMin: "12GB RAM",
    ollamaName: "mistral:7b",
    hfUrl: "https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3",
    useCases: JSON.stringify(["APIs", "agents", "classification"]),
  },
  {
    slug: "phi-3-mini",
    name: "Phi-3 Mini",
    vendor: "Microsoft",
    description:
      "Compact model optimized for reasoning and coding on constrained hardware. Ideal when you want local inference on lighter machines.",
    sizeParams: "3.8B",
    license: "MIT",
    contextLength: 128000,
    tags: JSON.stringify(["coding", "reasoning", "small"]),
    hardwareMin: "8GB RAM",
    ollamaName: "phi3:mini",
    hfUrl: "https://huggingface.co/microsoft/Phi-3-mini-4k-instruct",
    useCases: JSON.stringify(["edge devices", "fast prototyping", "coding"]),
  },
  {
    slug: "qwen2-5-coder-7b",
    name: "Qwen2.5 Coder 7B",
    vendor: "Alibaba",
    description:
      "Code-specialized open model that shines at completion, refactoring, and explaining repositories when run locally with a code UI.",
    sizeParams: "7B",
    license: "Apache 2.0",
    contextLength: 131072,
    tags: JSON.stringify(["coding", "completion", "repo"]),
    hardwareMin: "16GB RAM",
    ollamaName: "qwen2.5-coder:7b",
    hfUrl: "https://huggingface.co/Qwen/Qwen2.5-Coder-7B-Instruct",
    useCases: JSON.stringify(["IDE plugins", "code review", "tests"]),
  },
  {
    slug: "gemma-2-9b",
    name: "Gemma 2 9B",
    vendor: "Google",
    description:
      "Balanced open-weight model from Google DeepMind. Strong multilingual chat and instruction following for local experimentation.",
    sizeParams: "9B",
    license: "Gemma Terms",
    contextLength: 8192,
    tags: JSON.stringify(["chat", "multilingual", "general"]),
    hardwareMin: "16GB RAM + GPU/Metal",
    ollamaName: "gemma2:9b",
    hfUrl: "https://huggingface.co/google/gemma-2-9b-it",
    useCases: JSON.stringify(["chat", "translation", "writing"]),
  },
  {
    slug: "deepseek-coder-v2-lite",
    name: "DeepSeek Coder V2 Lite",
    vendor: "DeepSeek",
    description:
      "MoE-style coding model that punches above its active parameter count. Useful for local coding agents and structured generation.",
    sizeParams: "16B MoE",
    license: "DeepSeek License",
    contextLength: 128000,
    tags: JSON.stringify(["coding", "agents", "moe"]),
    hardwareMin: "24GB RAM",
    ollamaName: "deepseek-coder-v2:lite",
    hfUrl: "https://huggingface.co/deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct",
    useCases: JSON.stringify(["agents", "tool calling", "codegen"]),
  },
  {
    slug: "nomic-embed-text",
    name: "Nomic Embed Text",
    vendor: "Nomic",
    description:
      "Open embedding model for local RAG pipelines. Pair with a vector store to keep retrieval fully on-device.",
    sizeParams: "137M",
    license: "Apache 2.0",
    contextLength: 8192,
    tags: JSON.stringify(["embeddings", "rag", "retrieval"]),
    hardwareMin: "4GB RAM",
    ollamaName: "nomic-embed-text",
    hfUrl: "https://huggingface.co/nomic-ai/nomic-embed-text-v1.5",
    useCases: JSON.stringify(["RAG", "semantic search", "clustering"]),
  },
  {
    slug: "llava-1-6",
    name: "LLaVA 1.6",
    vendor: "LLaVA Team",
    description:
      "Vision-language model for local image understanding. Useful for screenshot analysis and multimodal prototypes without cloud APIs.",
    sizeParams: "7B–34B",
    license: "Apache 2.0",
    contextLength: 4096,
    tags: JSON.stringify(["vision", "multimodal", "images"]),
    hardwareMin: "16GB RAM + GPU",
    ollamaName: "llava:7b",
    hfUrl: "https://huggingface.co/llava-hf/llava-v1.6-mistral-7b-hf",
    useCases: JSON.stringify(["screenshots", "docs OCR assist", "multimodal chat"]),
  },
];

const tools = [
  {
    slug: "ollama",
    name: "Ollama",
    category: "runtime",
    description:
      "The fastest on-ramp to local LLMs. Pull models, run an OpenAI-compatible API on localhost, and script against it from any language.",
    website: "https://ollama.com",
    github: "https://github.com/ollama/ollama",
    platforms: JSON.stringify(["macOS", "Linux", "Windows"]),
    tags: JSON.stringify(["api", "cli", "beginner", "openai-compatible"]),
    difficulty: "beginner",
  },
  {
    slug: "lm-studio",
    name: "LM Studio",
    category: "ui",
    description:
      "Desktop app for discovering GGUF models, chatting locally, and exposing a local server. Ideal when you want a GUI plus an API.",
    website: "https://lmstudio.ai",
    github: null,
    platforms: JSON.stringify(["macOS", "Linux", "Windows"]),
    tags: JSON.stringify(["gui", "gguf", "openai-compatible"]),
    difficulty: "beginner",
  },
  {
    slug: "llama-cpp",
    name: "llama.cpp",
    category: "runtime",
    description:
      "High-performance C/C++ inference for GGUF models. The foundation behind many local tools and the gold standard for CPU/Metal/CUDA efficiency.",
    website: "https://github.com/ggerganov/llama.cpp",
    github: "https://github.com/ggerganov/llama.cpp",
    platforms: JSON.stringify(["macOS", "Linux", "Windows"]),
    tags: JSON.stringify(["gguf", "performance", "cli"]),
    difficulty: "intermediate",
  },
  {
    slug: "vllm",
    name: "vLLM",
    category: "runtime",
    description:
      "Throughput-oriented inference server for GPU machines. Use it when you need production-like local serving with continuous batching.",
    website: "https://docs.vllm.ai",
    github: "https://github.com/vllm-project/vllm",
    platforms: JSON.stringify(["Linux"]),
    tags: JSON.stringify(["gpu", "serving", "openai-compatible"]),
    difficulty: "advanced",
  },
  {
    slug: "open-webui",
    name: "Open WebUI",
    category: "ui",
    description:
      "Self-hosted ChatGPT-style interface that talks to Ollama and other backends. Great for teams exploring local AI with a familiar UX.",
    website: "https://openwebui.com",
    github: "https://github.com/open-webui/open-webui",
    platforms: JSON.stringify(["Docker", "macOS", "Linux", "Windows"]),
    tags: JSON.stringify(["chat-ui", "self-hosted", "ollama"]),
    difficulty: "beginner",
  },
  {
    slug: "langchain",
    name: "LangChain",
    category: "framework",
    description:
      "Composable framework for chains, agents, and RAG. Point it at a local OpenAI-compatible endpoint and keep your stack portable.",
    website: "https://js.langchain.com",
    github: "https://github.com/langchain-ai/langchainjs",
    platforms: JSON.stringify(["Node", "Python"]),
    tags: JSON.stringify(["rag", "agents", "framework"]),
    difficulty: "intermediate",
  },
  {
    slug: "llamaindex",
    name: "LlamaIndex",
    category: "framework",
    description:
      "Data framework focused on indexing documents and building retrieval pipelines. Pair with local embeddings for private knowledge bases.",
    website: "https://www.llamaindex.ai",
    github: "https://github.com/run-llama/llama_index",
    platforms: JSON.stringify(["Python", "TypeScript"]),
    tags: JSON.stringify(["rag", "indexing", "framework"]),
    difficulty: "intermediate",
  },
  {
    slug: "continue-dev",
    name: "Continue",
    category: "ide",
    description:
      "Open-source coding assistant for VS Code and JetBrains that can target local models via Ollama or LM Studio.",
    website: "https://continue.dev",
    github: "https://github.com/continuedev/continue",
    platforms: JSON.stringify(["VS Code", "JetBrains"]),
    tags: JSON.stringify(["ide", "coding", "ollama"]),
    difficulty: "beginner",
  },
  {
    slug: "anythingllm",
    name: "AnythingLLM",
    category: "ui",
    description:
      "All-in-one desktop/server app for private RAG with local models, document workspaces, and agent tools.",
    website: "https://anythingllm.com",
    github: "https://github.com/Mintplex-Labs/anything-llm",
    platforms: JSON.stringify(["macOS", "Linux", "Windows", "Docker"]),
    tags: JSON.stringify(["rag", "desktop", "agents"]),
    difficulty: "beginner",
  },
  {
    slug: "huggingface-tgi",
    name: "Text Generation Inference",
    category: "runtime",
    description:
      "Hugging Face's production inference toolkit. Useful when you want optimized serving of Transformers models on your own GPUs.",
    website: "https://huggingface.co/docs/text-generation-inference",
    github: "https://github.com/huggingface/text-generation-inference",
    platforms: JSON.stringify(["Linux", "Docker"]),
    tags: JSON.stringify(["serving", "gpu", "transformers"]),
    difficulty: "advanced",
  },
];

const guides = [
  {
    slug: "what-is-local-ai",
    title: "What is local AI?",
    summary:
      "Understand why developers run models on their own machines—and what tradeoffs come with privacy, cost, and latency.",
    level: "beginner",
    sortOrder: 1,
    readingMins: 6,
    content: `# What is local AI?

Local AI means running language, vision, or embedding models on hardware you control—your laptop, a home GPU box, or a team workstation—instead of calling a hosted API.

## Why developers care

- **Privacy**: source code, customer data, and notes never leave the machine.
- **Cost control**: no per-token bill while you iterate.
- **Latency**: local loops feel snappy for IDE tooling and agents.
- **Offline**: trains you to build resilient apps that work without the cloud.

## What "local" usually looks like

1. You download a model weights file (often GGUF or safetensors).
2. A runtime (Ollama, llama.cpp, vLLM, LM Studio) loads it.
3. You chat in a UI or hit an HTTP API that mimics OpenAI's shape.
4. Your app talks to \`localhost\` the same way it would talk to a cloud provider.

## Tradeoffs

Local models are smaller than frontier APIs. Quality depends on VRAM/RAM, quantization, and prompting. Start small (3B–8B), measure, then scale up.

## Next step

Install a runtime, pull one chat model and one embedding model, then build a tiny script that hits the local OpenAI-compatible endpoint.`,
  },
  {
    slug: "first-local-stack",
    title: "Your first local stack in 15 minutes",
    summary:
      "Install Ollama, pull Llama 3.1, and call it from curl and Node—the practical on-ramp for most developers.",
    level: "beginner",
    sortOrder: 2,
    readingMins: 8,
    content: `# Your first local stack in 15 minutes

This guide gets you from zero to a working local API.

## 1. Install Ollama

Download from [ollama.com](https://ollama.com) for macOS, Linux, or Windows. Confirm the CLI works:

\`\`\`bash
ollama --version
\`\`\`

## 2. Pull a chat model

\`\`\`bash
ollama pull llama3.1:8b
\`\`\`

Then try an interactive session:

\`\`\`bash
ollama run llama3.1:8b
\`\`\`

## 3. Hit the local HTTP API

Ollama exposes an OpenAI-compatible surface on \`http://localhost:11434\`.

\`\`\`bash
curl http://localhost:11434/api/generate -d '{
  "model": "llama3.1:8b",
  "prompt": "Explain REST APIs in two sentences.",
  "stream": false
}'
\`\`\`

## 4. Call it from Node

\`\`\`js
const res = await fetch("http://localhost:11434/api/chat", {
  method: "POST",
  headers: { "Content-Type": "application/json" },
  body: JSON.stringify({
    model: "llama3.1:8b",
    messages: [{ role: "user", content: "Say hello as a local model." }],
    stream: false,
  }),
});
console.log(await res.json());
\`\`\`

## 5. Add embeddings for RAG later

\`\`\`bash
ollama pull nomic-embed-text
\`\`\`

You now have chat + embeddings running entirely on your machine.`,
  },
  {
    slug: "models-vs-tools",
    title: "Models vs tools: how to choose",
    summary:
      "A decision guide for picking runtimes, UIs, frameworks, and model sizes based on your hardware and goals.",
    level: "beginner",
    sortOrder: 3,
    readingMins: 7,
    content: `# Models vs tools: how to choose

Local AI has two layers: **models** (the weights) and **tools** (everything that loads, serves, or wraps them).

## Models

Pick by task and hardware:

| Goal | Start with |
|------|------------|
| General chat | Llama 3.1 8B, Gemma 2 9B |
| Coding | Qwen2.5 Coder, DeepSeek Coder |
| Tiny / CPU | Phi-3 Mini |
| RAG embeddings | Nomic Embed Text |
| Images | LLaVA |

Quantization (Q4/Q5 GGUF) trades a little quality for much less memory.

## Tools

- **Need speed to first token of joy?** Ollama or LM Studio.
- **Need max performance on GPU?** vLLM or llama.cpp tuned builds.
- **Need a ChatGPT-like UI?** Open WebUI or AnythingLLM.
- **Building app logic?** LangChain / LlamaIndex against a local endpoint.
- **Coding in the IDE?** Continue pointed at Ollama.

## Rule of thumb

Keep the **interface contract** stable (OpenAI-compatible HTTP). Swap models and runtimes underneath without rewriting your product code.`,
  },
  {
    slug: "local-rag-basics",
    title: "Local RAG without the cloud",
    summary:
      "Chunk docs, embed locally, store vectors, and answer questions with a local LLM—all on localhost.",
    level: "intermediate",
    sortOrder: 4,
    readingMins: 10,
    content: `# Local RAG without the cloud

Retrieval-Augmented Generation (RAG) lets a small local model answer questions about *your* files.

## Pipeline

1. **Ingest** markdown, PDFs, or code.
2. **Chunk** into ~400–800 token pieces with overlap.
3. **Embed** each chunk with a local embedding model.
4. **Store** vectors (SQLite+sqlite-vss, Chroma, LanceDB, Qdrant).
5. **Retrieve** top-k chunks for a query.
6. **Generate** an answer with a local chat model using those chunks as context.

## Minimal mental model

\`\`\`
query → embed → nearest chunks → prompt(context + question) → local LLM
\`\`\`

## Practical tips

- Keep prompts explicit: "Answer only from context. If missing, say you don't know."
- Start with 3–5 retrieved chunks; more context is not always better.
- Re-embed when docs change.
- Log retrieved chunks during debugging—most RAG bugs are retrieval bugs.

## Localbench path

Favorite an embedding model (Nomic) and a chat model (Llama/Mistral), then pick a framework tool (LlamaIndex or LangChain) from Explore.`,
  },
  {
    slug: "openai-compatible-local-apis",
    title: "Design apps against local OpenAI-compatible APIs",
    summary:
      "Write backend code once, point it at localhost during development, and keep the door open to cloud providers later.",
    level: "intermediate",
    sortOrder: 5,
    readingMins: 9,
    content: `# Design apps against local OpenAI-compatible APIs

Most local runtimes speak a dialect of the OpenAI Chat Completions API. Treat that as your app's boundary.

## Pattern

\`\`\`ts
const baseURL = process.env.LLM_BASE_URL ?? "http://localhost:11434/v1";
const apiKey = process.env.LLM_API_KEY ?? "ollama";

export async function chat(messages: { role: string; content: string }[]) {
  const res = await fetch(\`\${baseURL}/chat/completions\`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: \`Bearer \${apiKey}\`,
    },
    body: JSON.stringify({
      model: process.env.LLM_MODEL ?? "llama3.1:8b",
      messages,
      temperature: 0.2,
    }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}
\`\`\`

## Why this matters

- Swap Ollama ↔ LM Studio ↔ vLLM by changing env vars.
- Promote to a hosted provider by changing \`baseURL\` and keys.
- Write integration tests against a deterministic local model.

## Backend checklist

- Timeouts and retries for cold model loads
- Streaming support for UX
- Separate embedding vs chat base URLs if needed
- Never hardcode secrets—even local keys belong in env

Localbench's own API layer follows the same idea: clear routes, JSON contracts, auth where it matters.`,
  },
  {
    slug: "hardware-and-quantization",
    title: "Hardware, VRAM, and quantization",
    summary:
      "Map model sizes to machines and learn how GGUF quantization makes local inference realistic.",
    level: "intermediate",
    sortOrder: 6,
    readingMins: 8,
    content: `# Hardware, VRAM, and quantization

## Rough sizing

| Machine | Comfortable models |
|---------|--------------------|
| 8GB unified / RAM | 3B–4B Q4 |
| 16GB | 7B–8B Q4/Q5 |
| 24GB VRAM | 13B–34B quantized / MoE lite |
| Multi-GPU server | full-precision or large MoE |

Rules are fuzzy—context length, batch size, and KV cache matter.

## Quantization in one paragraph

Quantization stores weights in fewer bits (e.g. 4-bit). GGUF Q4_K_M is a popular balance. You lose a little quality, gain a lot of speed/memory headroom.

## Developer workflow

1. Start with the smallest model that passes your eval prompts.
2. Measure tokens/sec and quality on *your* tasks.
3. Increase size only when quality plateaus.
4. Prefer more context + better retrieval over blindly larger models for RAG.

## Apple Silicon note

Metal acceleration in llama.cpp / Ollama makes Macs excellent for local AI. Prefer GGUF runtimes there.`,
  },
];

async function main() {
  await prisma.favorite.deleteMany();
  await prisma.user.deleteMany();
  await prisma.aiModel.deleteMany();
  await prisma.tool.deleteMany();
  await prisma.guide.deleteMany();

  await prisma.aiModel.createMany({ data: models });
  await prisma.tool.createMany({ data: tools });
  await prisma.guide.createMany({ data: guides });

  const passwordHash = await hash("demo1234", 10);
  await prisma.user.create({
    data: {
      email: "dev@localbench.dev",
      username: "demo",
      displayName: "Demo Developer",
      passwordHash,
    },
  });

  console.log(
    `Seeded ${models.length} models, ${tools.length} tools, ${guides.length} guides, and demo user (dev@localbench.dev / demo1234)`,
  );
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
